osvr/RenderKit/RenderManagerBase.cpp

```diff
@@ -1419,206 +1446,271 @@ namespace renderkit {
+		double overlapFrac =
+			m_params.m_displayConfiguration->getOverlapPercent();
+		if (overlapFrac < 1.) {
+			const auto hfov =
+				m_params.m_displayConfiguration->getHorizontalFOV();
+			const auto angularOverlap = hfov * overlapFrac;
+			rotateEyesApart = util::getDegrees((hfov - angularOverlap) / 2.);
+		}
+		// Right eyes should rotate the other way.
+		if (whichEye % 2 != 0) {
+			rotateEyesApart *= -1;
+		}
+		rotateEyesApart = osvr::common::degreesToRadians(rotateEyesApart);
+		q_from_axis_angle(q_rotatedEyeFromEye.quat, 0, 1, 0, rotateEyesApart);
+
+		/// Include the impact of the eyeFromHead matrix.
+		// This is a translation along the X axis in head space by
+		// the IPD, or its negation, depending on the eye.
+		// We assume that even eyes are left eyes and odd eyes are
+		// right eyes.  We further assume that head space is between
+		// the two eyes.  If the display descriptor wants us to swap
+		// eyes, we do so by inverting the offset for each eye.
+		/// @todo dealing with eye tracking here or mono displays
+		q_xyz_quat_type q_headFromRotatedEye;
+		makeIdentity(q_headFromRotatedEye);
+		q_xyz_quat_type q_roomFromEye;
+		OSVR_TimeValue timestamp;
+
+		// If eye poses are present, use them instead of head +- IPD
+		bool hasEyes = hasLeftEyePose() && hasRightEyePose();
+		if (!hasEyes)
+		{
+			if (whichEye % 2 == 0) {
+				// Left eye
+				q_headFromRotatedEye.xyz[Q_X] -= params.IPDMeters / 2;
+			}
+			else {
+				// Right eye
+				q_headFromRotatedEye.xyz[Q_X] += params.IPDMeters / 2;
+			}
+		}
+		else
+		{
+			if (whichEye % 2 == 0) {
+				// Left eye
+				if (params.roomFromLeftEyeReplace != nullptr) {
+					q_from_OSVR(q_roomFromEye, *params.roomFromLeftEyeReplace);
+				}
+				else
+				{
+					/// Use the state interface to read the most-recent
+					/// location of the left eye.  It will have been updated
+					/// by the most-recent call to update() on the context.
+					/// DO NOT update the client here, so that we're using the
+					/// same state for all eyes.
+					if (!m_leftEyePoseCache || !m_leftEyePoseCache->getLastReport(timestamp, m_roomFromLeftEye)) {
+						// This it not an error -- they may have put in an invalid
+						// state name for the left eye; we just ignore that case.
+					}
+					q_headFromRotatedEye.xyz[Q_X] -= m_roomFromLeftEye.translation.data[0];
+					q_headFromRotatedEye.xyz[Q_Y] -= m_roomFromLeftEye.translation.data[1];
+					q_headFromRotatedEye.xyz[Q_Z] -= m_roomFromLeftEye.translation.data[2];
+
+				}
+			}
+			else {
+				/// Use the state interface to read the most-recent
+				/// location of the right eye.  It will have been updated
+				/// by the most-recent call to update() on the context.
+				/// DO NOT update the client here, so that we're using the
+				/// same state for all eyes.
+				if (!m_rightEyePoseCache || !m_rightEyePoseCache->getLastReport(timestamp, m_roomFromRightEye)) {
+					// This it not an error -- they may have put in an invalid
```

Comment:
Same comment here as in the left eye version above. Should we have a proper fallback in case of error? Is m_roomFromRightEye going to be valid data if this if statement falls through to the code below?